pkg/fqdn/rulegen.go

```diff
@@ -145,6 +145,47 @@ perRule:
 	}
 }
 
+// AreFQDNRulesStillPresent is a function that validates that the rule UUID is
+// still present in the FQDN rules, it returns true if the rules do not have
+// FQDN rule or are still in the repo, returns false if any rule were deleted
+// from gen.AllRules.
+// The main reason is to avoid race conditions in the policyAdd, a new CNP
```

Comment:
I would put this explanation in the place we call this function instead. The actual race is in how all this is connected in daemon, specifically around the order of operations when different things call `PolicyAdd`. I would go into a little more detail (I'm assuming that the check is done with the lock held, though, so the text may need to change if you don't want to do that):
This check is needed to avoid a race where a DNS update to a rule enqueues an copy of that rule that is older than a k8s/API update to that same rule. Rules are assigned random UUIDs in MarkToFQDNsRule when initially inserted and DNS updates always replace existing rules. In this out-of-order scenario the k8s/API updated rule would receive a new UUID and the older UUID, belonging to the DNS updated rule, would no longer be present. This check fails in that case, and relies on the Add/Remove of a toFQDNs rule in RuleGen happening atomically.